cmd2/cmd2.py

```diff
@@ -3123,8 +3185,134 @@ def cmdloop(self, intro: Optional[str]=None) -> None:
             self._cmdloop()
 
         # Run the postloop() no matter what
+        for func in self._postloop_hooks:
+            func()
         self.postloop()
 
+    ###
+    #
+    # plugin related functions
+    #
+    ###
+    def _initialize_plugin_system(self):
+        """Initialize the plugin system"""
+        self._preloop_hooks = []
+        self._postloop_hooks = []
+        self._postparsing_hooks = []
```

Comment:
It seems asymmetric that there is no ``self._preparsing_hooks``.  

Worse, ``preparse`` is currently never called.  The main reason that bug ticket #417 was created is that in the port from a parser based on ``pyparsing`` to one based on ``shlex`` in the 0.9.0 release, we accidentally stopped calling ``preparse``.  Even if we decide to deprecate it, it should be called now for backwards compatibility.

One possible place to call it would be at the beginning of ``_complete_statement``:
```Python
    def _complete_statement(self, line: str) -> Statement:
        """Keep accepting lines of input until the command is complete.

        There is some pretty hacky code here to handle some quirks of
        self.pseudo_raw_input(). It returns a literal 'eof' if the input
        pipe runs out. We can't refactor it because we need to retain
        backwards compatibility with the standard library version of cmd.
        """
        statement = self.statement_parser.parse(self.preparse(line))
```

Reply:
`preparse()` is now called by `_complete_statement()`. It is asymmetric that there are no `self._preparsing_hooks`. Here's why.

If it's a multiline command, `preparse()` may not get all the user input. `_complete_statement()` really does two things: a) parse the user input, and b) accept more input in case it's a multiline command the passed string doesn't have a terminator. If we call `preparse()` at the beginning (which is now what it does), we call `preparse()` before we parse the input, but if it's a multiline command and there is no terminator, `preparse()` won't have all the input because it hasn't been entered yet. That's undesirable and misleading.

If we call `preparse()` after we are sure we have all the input, then by definition it can't be before we have parsed the input, because we have to parse it to know if we have it all. It's also misleading to have a hook called `preparse()` which doesn't get called until after everything is parsed.

The good news is that parsing (finally) is fast. If a developer wants to change the user input, they can write a post parsing hook, which gets passed the complete user input, modify it as desired, and reparse it.

By not having a `self._preparsing_hooks` we just make it less confusing. Postparsing hooks have all the input, and can choose to modify it if necessary before the statement is executed.

Reply:
That sounds like a reasonable explanation to me for why it is asymmetric.